### PR TITLE
Add new `harvest` command + `harvester` device

### LINKED ISSUE #490: Give the player control over whether growable entities regrow
After some discussion in the comments, I am refocusing on the central issues:

- The player ought to be able to control when and where growing things regrow (except for instantly regrowing things like water).
- I think *not* regrowing ought to be the default.  If the player wants to farm things, they have to do a bit more work.

Below is my original proposal for a `harvest` command, though that is not the only possible solution (and indeed, probably not the best solution).  I would be happy with any solution that fulfills the above two criteria and does not complicate things too much.

Original `harvest` proposal
------------------------------------

**tl;dr**: I propose providing both a `grab` command and a `harvest` command.  The former would simply remove an entity from the current cell; the latter would leave behind a seed if the entity is growable (*i.e.* what the `grab` command currently does).

We could simply have both commands provided by a `grabber`; or, perhaps a better idea would be to have a `grabber` only provide `grab`, whereas to `harvest` you need to build a `harvester` device, which would have some suitably simple recipe (e.g. something you can make out of a few trees) since it would be one of the first things you want to build.

Motivation
--------------

- This would greatly simplify the process of clearing an area for farming.
- It turns out that having growable entities *always* regrow *every* time you grab them is limiting and annoying.  Read on for the full story of how I realized this below.

Backstory
--------------

I had an issue with my tree and bit harvesting operations.  It would work fine for a while but then the flow of trees or bits would slow to a trickle.  It took me an embarrassingly long time to figure out what was wrong, but with the help of #483 I was able to figure out which robot was failing, and then with some loggers I was able to find the exception that caused it to fail.  Once I knew that I figured out the rest pretty quickly.

As [described here](https://github.com/swarm-game/swarm/issues/393#issuecomment-1161845144), I have come up with a simple scheme to support pull-based manufacturing: a *provider* robot obtains some resource somehow, and then it sits on a particular square and places one entity.  When it notices the cell is empty, it places another entity, and so on.   Other robots that want to obtain the resource come to the square, wait for an entity to appear, and then grab it.  (Of course the grabbing should happen atomically, as in #479 , but that's not the issue here.)

My tree processing operation involves 5 robots, but for the purposes of this story you only need to know about three:
- A harvester robot continually goes over a rectangular plot of trees, harvesting any trees that grow, and giving them all to the tree provider robot.
- The tree provider robot sits in the same place and acts as a tree provider (in the sense of the description above).
- Another tree processing robot obtains trees from the tree provider, processes them into branches and logs, then gives the results to some other robots.

This worked fine for a while but it seemed to stop working eventually.  I suspected some kind of race condition but I couldn't figure it out. After we merged #483 , I was able to determine that eventually it was the tree provider that switched to `idle`, i.e. it had crashed, and trees just started accumulating in its inventory. Mysteriously, though, the tree processing robot still occasionally processed some trees.  Once I put a logger in the tree provider robot, I found out that `place` was crashing with the message `There is already an entity here`.  But I didn't know how that could be possible --- the tree provider was the *only* robot that ever `place`d anything in that cell.  The tree harvester would simply `give` its trees to the tree provider, and the tree processor would only ever `grab` trees placed by the provider.

Have you figured it out yet? :smile: 

It suddenly dawned on me that every time the tree processing robot `grab`bed a tree from the cell of the tree provider, a seed started to grow!  In fact, after a while, there would in theory be a whole bunch of tree seed robots all growing in that same cell.  It was just a matter of time before encountering the standard `grab`/`place` race condition (#479) --- the tree provider would be interrupted just after testing the cell and finding it empty; a tree seed would then finish growing; and hence, from the tree provider's point of view, a tree grew in the instant in between checking to make sure the cell was empty and executing `place`.  The tree processor kept working since trees still kept showing up in the provider cell.  Probably even more than one were growing concurrently, so it was more than we would get from harvesting a single tree, but the overall effect was a much reduced throughput.

So part of the solution could of course be a `placeif` primitive like in #479 .  That would prevent the tree provider from crashing in this case.  But we also just really don't want trees to regrow in the tree provider cell. At that point, we want to just think of them as a commodity to be passed around, not as something that should grow.

Lessons learned
----------------------

- Shared memory concurrency is hard.
- Good debugging tools are worth their weight in `mithril`.

### PR DESCRIPTION
**Pull request by byorgey**
Make a new `harvester` device + `harvest` command.  The recipe for a `harvester` is 2 `board` + 2 `wooden gear` + 1 `box`, hence it can be made completely out of 3 `tree`s. (3 trees make 12 `board`s; 6 `board`s are for the `box`, 4 are to make the 2 `wooden gear`s).

The `harvester` device provides the `harvest` and `grab` capabilities, hence it supports both the `harvest` and `grab` commands. 
The `harvest` command now does exactly what the `grab` command used to do; the `grab` command works identically except that it never leaves behind a growing seed. 

As an aside, I am actually not sure what happens if you try to `build` a robot that uses both `grab` and `harvest` (but none of the other capabilities provided by a `grabber`).  Is the capability analysis smart enough to realize that it should just install a `harvester`, or might it end up installing both a `harvester` and a `grabber` unnecessarily?

Closes #490 (I split out another part of that issue into #502).  @xsebek , I know in #490 you expressed an objection to adding a `harvester` device so I'm curious what you think about this.

### CODE REVIEW

**Comment by byorgey** [comment]
I'm not 100% sure why the test doesn't work, but it seems to have something to do with the `teleport` command.  I think maybe what is happening is that the program to check the win condition is run as a robot with ID 0, but there is already a robot with ID 0 in the robot map, so it gets confused when trying to do `teleport self`.

**Comment by byorgey** [comment]
I fixed it by using plain old `move` instead of `teleport`, but we should figure out what's going on.

File: data/scenarios/Testing/harvest.yaml [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    '┌': [stone, upper left corner]
+    '┐': [stone, upper right corner]
+    '└': [stone, lower left corner]
+    '┘': [stone, lower right corner]
+    '─': [stone, horizontal wall]
+    '│': [stone, vertical wall]
+  upperleft: [-1, 1]
+  map: |
+    ┌───┐
+    │.TT│
```
Comment (xsebek) [inline data/scenarios/Testing/harvest.yaml:35]: Maybe add another row for water?

**Comment by xsebek** [comment]
@byorgey looks nice, I might yet take a liking to the harvester. :+1: 

> I am actually not sure what happens if you try to build a robot that uses both grab and harvest...

Well, let's write a test and if it does not work, we can file a bug. :grinning: <sub>I bet it is not smart enough. :thinking: </sub>

File: data/entities.yaml [hunk 2]
```diff
@@ -598,6 +598,21 @@
...
   properties: [portable]
 
+- name: harvester
+  display:
+    attr: device
+    char: '≤'
+  description:
+  - A harvester can be used via the 'harvest' command, which is
+    almost identical to the 'grab' command.  The big difference
+    is that some entities, when harvested instead of grabbed,
```
[the remark this answers is not in the document]
Comment (TristanCacqueray) [reply data/entities.yaml:608]: I guess we should display such properties in the info panel.

File: data/scenarios/Testing/harvest.yaml [hunk 3]
```diff
@@ -0,0 +1,36 @@
+name: Test harvest command
+description: Test the difference between the grab and harvest commands.
```
Comment (xsebek) [inline data/scenarios/Testing/harvest.yaml:2]:
```suggestion
description: |
  Test the difference between the grab and harvest commands.
  https://github.com/swarm-game/swarm/issues/490
```


I admit I like being able to click on the PR link in the terminal:
Screenshot from 2022-06-29 23-22-28

File: src/Swarm/Game/Step.hs [hunk 4]
```diff
@@ -1598,6 +1563,46 @@ execConst c vs s k = do
...
+    -- Ensure it can be picked up.
+    (e `hasProperty` Portable)
+      `holdsOrFail` ["The", e ^. entityName, "here can't be grabbed."]
+
+    -- Remove the entity from the world.
+    updateEntityAt loc (const Nothing)
+    flagRedraw
+
+    -- Possibly regrow the entity.
+    when ((e `hasProperty` Growable) && shouldHarvest) $ do
```
[the remark this answers is not in the document]
Comment (byorgey) [reply src/Swarm/Game/Step.hs:1582]: Hah, oops! :facepalm: Will fix and add to the test.

File: data/entities.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (byorgey) [reply data/entities.yaml:608]: Sure, displaying properties like growable specially in the info panel sounds good to me.

File: src/Swarm/Game/Step.hs (hunk 4 above)
[the remark this answers is not in the document]
Comment (byorgey) [reply src/Swarm/Game/Step.hs:1587]: Done.  Made a new `infinite` property which makes more sense than making `growth: [0,0]` mean something special, anyway.

File: data/entities.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (byorgey) [reply data/entities.yaml:605]: @xsebek , @TristanCacqueray what do you think of it now?  I kind of threw something together, happy for suggestions to make it better.

**Comment by byorgey** [comment]
> Well, let's write a test and if it does not work, we can file a bug. :grinning: I bet it is not smart enough. :thinking:

Sure enough, you are right.  When I do `build {grab; harvest}` it ends up installing both a `grabber` and a `harvester`.  I filed #508 to track this.

**Comment by byorgey** [comment]
@xsebek thanks for the great ideas, comments, and catching a bug!  I think this is again ready for review now.

File: data/scenarios/Testing/490-harvest.yaml [hunk 5]
```diff
@@ -1,15 +1,19 @@
...
     t1 <- ishere "tree";
     move;
     t2 <- ishere "tree";
-    return (t1 && not t2)
+    t3 <- as base {n <- count "water"; return (n == 4)};
+    return (t1 && not t2 && t3)
   } { return false }
 solution: |
-  move; harvest; move; grab
+  move; harvest; move; grab; turn right; move; grab; grab; grab; grab
```
Comment (xsebek) [inline data/scenarios/Testing/490-harvest.yaml:16]: Could you make one of those a harvest?

I expect it makes no difference, but I would like to have the expected behaviour tested.

**Review by xsebek — APPROVED** [review]
:house_with_garden: :farmer: 

I have some nitpicks, but all the important things are already addressed.

File: src/Swarm/TUI/View.hs [hunk 6]
```diff
@@ -733,8 +733,27 @@ explainFocusedItem s = case focusedItem s of
...
+    ]
+
+displayProperties :: [EntityProperty] -> Widget Name
+displayProperties = displayList . mapMaybe showProperty
+ where
+  showProperty Growable = Just "growing"
+  showProperty Infinite = Just "infinite"
+  showProperty Liquid = Just "liquid"
+  showProperty Unwalkable = Just "blocking"
+  showProperty _ = Nothing
```
[the remark this answers is not in the document]
Comment (byorgey) [reply src/Swarm/TUI/View.hs]: Good idea, done.

File: src/Swarm/TUI/View.hs [hunk 7]
```diff
@@ -733,8 +733,27 @@ explainFocusedItem s = case focusedItem s of
 
 explainEntry :: AppState -> Entity -> Widget Name
 explainEntry s e =
-  displayParagraphs (e ^. entityDescription)
-    <=> explainRecipes s e
+  vBox
+    [ displayProperties (e ^. entityProperties)
```
[the remark this answers is not in the document]
Comment (byorgey) [reply src/Swarm/TUI/View.hs:737]: Sure, being able to filter/sort the inventory in general would be nice.  Right now it is just in alphabetical order.  Actually, it's in ASCII order, so all uppercase letters come first...

[the remark this answers is not in the document]
Comment (byorgey) [reply src/Swarm/TUI/View.hs:737]: Filed #519 .

**Comment by byorgey** [comment]
Wait a minute, CI is failing because one of the tests we expected to fail is actually succeeding...

**Comment by byorgey** [comment]
Oh, at least I now figured out why CI was failing here but succeeding for me locally: it's because here it is running CI on a branch with `main` merged in.  When I merge `main` into this branch locally, CI starts failing for me too.

**Comment by byorgey** [comment]
I don't understand what's going on with #508 , but I don't think it needs to hold up this PR any more!
